### PR TITLE
Report warning using `foreach` with nullable value

### LINKED ISSUE #23493: Report warning using foreach with a nullable value
A warning should be reported using `foreach` with a nullable value.
```
class C
{
    static void F(object[]? args)
    {
        foreach (var o in args) // warn
        {
        }
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #23493.

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -34470,6 +34470,232 @@ static void F(A<object?>[] c)
...
+                // (13,27): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(IEnumerable)t1").WithLocation(13, 27),
+                // (13,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable)t1").WithLocation(13, 27));
+        }
+
+        [Fact]
+        public void ForEach_15()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >ForEach_15 [](start = 20, length = 10)

Some more test ideas:
- `MyEnumerator GetEnumerator()` where `MyEnumerator.Current` returns a nullable type, but the element type of the `foreach` is non-nullable
- `MyEnumerator GetEnumerator()` where `MyEnumerator.Current` returns a non-nullable type which has a user-defined conversion to the non-nullable element type of the `foreach`, but that user-defined conversion returns a nullable

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: `ForEach_02` tests `object? Current { get; }`. Added a test with a user-defined conversion.

---
In reply to: [214110121](https://github.com/dotnet/roslyn/pull/29587#discussion_r214110121) [](ancestors = 214110121)

**Comment by cston** [comment]
@dotnet/roslyn-compiler for a second review, thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -34470,6 +34470,269 @@ static void F(A<object?>[] c)
...
+            x.ToString();
+        foreach (B y in e)
+            y.ToString();
+        foreach (B? z in e)
+            z.ToString();
+    }
+}";
+            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
+            comp.VerifyDiagnostics(
+                // (15,18): warning CS8601: Possible null reference assignment.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This error message is especially unhelpful in this context, as there's no assignment happening. Should we add a better message?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -34470,6 +34470,269 @@ static void F(A<object?>[] c)
...
+                // (13,27): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(IEnumerable)t1").WithLocation(13, 27),
+                // (13,27): warning CS8602: Possible dereference of a null reference.
+                //         foreach (var v in (IEnumerable)t1) // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "(IEnumerable)t1").WithLocation(13, 27));
+        }
+
+        [Fact]
+        public void ForEach_16()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Another test idea: check the foreach expression for null before use.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Added comment.

---
In reply to: [214179736](https://github.com/dotnet/roslyn/pull/29587#discussion_r214179736) [](ancestors = 214179736)

**Comment by cston** [comment]
@dotnet-bot test this

**Pull request merged**